sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/ZPageServer.java

```diff
@@ -0,0 +1,165 @@
+ *     ZPageServer.registerAllPagesToHttpServer(server);
+ *     server.start();
+ *     ... // do work
+ *   }
+ * }
+ * }</pre>
+ */
+@ThreadSafe
+public final class ZPageServer {
+  // The maximum number of queued incoming connections allowed on the HttpServer listening socket.
+  private static final int HTTPSERVER_BACKLOG = 5;
+  // Length of time to wait for the HttpServer to stop
+  private static final int HTTPSERVER_STOP_DELAY = 1;
+  // Tracez SpanProcessor and DataAggregator for constructing TracezZPageHandler
+  private static final TracezSpanProcessor tracezSpanProcessor =
+      TracezSpanProcessor.newBuilder().build();
+  private static final TracezDataAggregator tracezDataAggregator =
+      new TracezDataAggregator(tracezSpanProcessor);
+  // Handler for /tracez page
+  private static final ZPageHandler tracezZPageHandler =
+      new TracezZPageHandler(tracezDataAggregator);
+
+  private static final Object mutex = new Object();
+  private static volatile AtomicBoolean isTracezSpanProcesserAdded = new AtomicBoolean(false);
+
+  @GuardedBy("mutex")
+  @Nullable
+  private static HttpServer server;
+
+  /** Function that adds the {@link TracezSpanProcessor} to the {@link tracerSdkProvider}. */
+  private static void addTracezSpanProcessor() {
+    if (isTracezSpanProcesserAdded.compareAndSet(false, true)) {
+      synchronized (mutex) {
+        TracerSdkProvider tracerProvider = OpenTelemetrySdk.getTracerProvider();
+        tracerProvider.addSpanProcessor(tracezSpanProcessor);
+      }
+    }
+  }
+
+  /**
+   * Returns a {@code ZPageHandler} for tracing debug. The page displays information about all
+   * running spans and all sampled spans based on latency and error.
+   *
+   * <p>It displays a summary table which contains one row for each span name and data about number
+   * of running and sampled spans.
+   *
+   * <p>Clicking on a cell in the table with a number that is greater than zero will display
+   * detailed information about that span.
+   *
+   * <p>This method will add the TracezSpanProcessor to the tracerProvider, it should only be called
+   * once.
+   *
+   * @return a {@code ZPageHandler} for tracing debug.
+   */
+  public static ZPageHandler getTracezZPageHandler() {
+    addTracezSpanProcessor();
+    return tracezZPageHandler;
+  }
+
+  /**
+   * Registers all zPages to the given {@link HttpServer} {@code server}.
+   *
+   * @param server the server that exports the zPages.
+   */
+  public static void registerAllPagesToHttpServer(HttpServer server) {
+    // For future zPages, register them to the server in here
+    server.createContext(
+        tracezZPageHandler.getUrlPath(), new ZPageHttpHandler(getTracezZPageHandler()));
+  }
+
+  /** Method for stopping the {@link HttpServer} {@code server}. */
+  private static void stop() {
+    synchronized (mutex) {
+      checkState(server != null, "The HttpServer is already stopped.");
```

Comment:
is it really worth throwing an exception if two threads try to stop the server? Seems like it should just be a no-op.

Reply:
Changed to no-op